src/details/ArborX_PairIndexRank.hpp

```diff
@@ -0,0 +1,41 @@
+/****************************************************************************
+ * Copyright (c) 2017-2023 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_PAIR_INDEX_RANK_HPP
+#define ARBORX_PAIR_INDEX_RANK_HPP
+
+#include <Kokkos_Macros.hpp>
+
+namespace ArborX
+{
+
+struct PairIndexRank
+{
+  int index;
+  int rank;
+
+private:
+  friend KOKKOS_FUNCTION constexpr bool operator==(PairIndexRank lhs,
+                                                   PairIndexRank rhs)
+  {
+    return lhs.index == rhs.index && lhs.rank == rhs.rank;
+  }
+  friend KOKKOS_FUNCTION constexpr bool operator<(PairIndexRank lhs,
+                                                  PairIndexRank rhs)
+  {
+    return lhs.index < rhs.index ||
+           (lhs.index == rhs.index && lhs.rank < rhs.rank);
+  }
```

Comment:
This seems like a bad default that we were using. It would make more sense to sort indices by ranks, then rank by indices. I suggest doing
```suggestion
  friend KOKKOS_FUNCTION constexpr bool operator<(PairIndexRank lhs,
                                                  PairIndexRank rhs)
  {
    return lhs.rank < rhs.rank ||
           (lhs.rank == rhs.rank && lhs.index < rhs.index);
  }
```
Thoughts?

I'm less certain about whether it should be `(index, rank)` or `(rank, index)` so not proposing to change that at the moment.